sycl/include/CL/sycl/detail/pi.h

```diff
@@ -1236,6 +1246,22 @@ __SYCL_EXPORT pi_result piKernelSetExecInfo(pi_kernel kernel,
                                             size_t param_value_size,
                                             const void *param_value);
 
+/// Creates PI kernel object from a native handle.
+/// NOTE: The created PI object takes ownership of the native handle.
```

Comment:
I think we should add new stuff with the possibility of controlling the ownership from the beginning. Level-Zero doesn't maintain reference counting for any objects, so it will be needed for all objects eventually that users have the control.

Reply:
Will you add the "ownership" argument here? I don't see how we could have "pi_native_handle" to contain that as that would require allocating a custom structure with unclear lifetime (who allocates/frees it and when).

Reply:
@smaslov-intel all these `make_*` functions accept native type by const reference. If user passes the custom object, I'd assume, that we can get address of that object, and use a pointer as `pi_native_handle`. This pointer will be valid for the duration of `make_*` call, and I can't come up with a situation, where we need a longer lifetime.

Reply:
I see where you are going, and agree this is doable. But that would sacrifice clarity and robustness, I believe. Basically details of the communication would be hidden by an opaque `pi_native_handle` whereas everything is explicit with real new arguments. I think I still prefer the explicit way, as this way we make the protocol/ABI change explicit too, which is a good thing. Also, we did so for "context" already:

https://github.com/intel/llvm/blob/ac54a4a765255d0e847d0a7663b881a054d7391b/sycl/include/CL/sycl/detail/pi.h#L990

Reply:
Ok, done